### PR TITLE
[BEAM-3601] Switch to Java 8 futures

### PR DESCRIPTION
**Pull request by kennknowles**
This ports the Beam codebase from Guava futures to Java 8 futures. This has been discussed at length on the Beam dev list.

The standards that I have held:

 - Consumers of futures use `CompletionStage`. They can do async chaining and can use `MoreFutures.get(...)` to get the result without passing through types they shouldn't be touching.
 - Producers of futures use `CompletableFuture` and return it as a `CompletionStage`.
 - When you need to handle exceptions in a reasonable way, use `MoreFutures` helpers that will complete things exceptionally based on the underlying function throwing.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand:
   - [x] What the pull request does
   - [x] Why it does it
   - [x] How it does it
   - [x] Why this approach
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

**Comment by kennknowles** [comment]
So many mocks ;_;

**Comment by kennknowles** [comment]
Requested reviews from just a couple folks in the right arenas rather than targeting the authors of every touched piece of code.

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java [hunk 1]
```diff
@@ -351,14 +352,14 @@ public DataflowPackage stageToFile(
...
       // TODO: Java 8 / Guava 23.0: FluentFuture
-      ListenableFuture<StagingResult> stagingResult =
-          Futures.transformAsync(
-              computePackageAttributes(sourcePackage, stagingPath),
-              packageAttributes -> stagePackage(packageAttributes, retrySleeper, createOptions));
-
-      ListenableFuture<DataflowPackage> stagedPackage =
-          Futures.transform(
-              stagingResult,
+      CompletionStage<StagingResult> stagingResult =
```
Comment (jkff) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java]: Remove the TODO above?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/MoreFutures.java [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+ *
+ * <ul>
+ *   <li>Always allow thrown exceptions, and they should cause futures to complete exceptionally.
+ *   <li>Always return {@link CompletionStage} as a future value.
+ *   <li>Return {@link CompletableFuture} only to the <i>producer</i> of a future value.
+ * </ul>
+ */
+public class MoreFutures {
+
+  /** Like {@link Runnable} but allowed to throw any exception. */
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/MoreFutures.java]: Clarify: are these interfaces intended to be used only in futures-related code? (if they are intended to be used more widely, they should be outside `MoreFutures`; if not, ideally that should be documented, cause it's tempting)

Comment (tgroh) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/util/MoreFutures.java]: In fact, these exist within the SDK harness, among other places, already, and I don't believe all of those uses interact directly with futures; I'd rather just have them in `util`, or `function`, (or any other meaningful name)

Comment (kennknowles) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/util/MoreFutures.java]: Done

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java (hunk 1 above)
Comment (kennknowles) [reply runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java]: Done

**Comment by kennknowles** [comment]
run java gradle precommit

**Pull request merged**